### PR TITLE
Refactor

### PR DESCRIPTION
**Pull request by edjamesmsft**
More refactoring work, mostly around initialization and message handling

### CODE REVIEW

File: Source/Services/Beam/beam_manager_impl.cpp [hunk 1]
```diff
@@ -57,11 +58,15 @@ beam_manager_impl::beam_manager_impl() :
...
     m_initGroupsComplete(false),
     m_initServerTimeComplete(false),
+    m_localUserInitialized(false),
     m_serverTimeOffset(std::chrono::milliseconds(0))
 {
     logger::create_logger();
     logger::get_logger()->set_log_level(log_level::error);
     logger::get_logger()->add_log_output(std::make_shared<debug_output>());
+
+    m_initializingTask = pplx::create_task([] {});
```
Comment (jicailiu) [inline Source/Services/Beam/beam_manager_impl.cpp]: >pplx::create_task([] {}); [](start = 25, length = 25)

= pplx::task_from_result(); is cheaper and take less binary space.

File: Source/Services/Beam/beam_manager_impl.cpp [hunk 2]
```diff
@@ -247,6 +275,88 @@ beam_manager_impl::init_worker(
...
+                        {
+                            process_reply(currentMessage->m_json);
+                        }
+                        else if (0 == messageType.compare(RPC_METHOD))
+                        {
+                            process_method(currentMessage->m_json);
+                        }
+                        else
+                        {
+                            LOGS_INFO << "Unexpected message from service";
```
Comment (jicailiu) [inline Source/Services/Beam/beam_manager_impl.cpp]: >LOGS_INFO [](start = 28, length = 9)

LOGS_* construct a stream, should use LOG_* if you don't need format the string.

File: Source/Services/Beam/beam_manager_impl.cpp [hunk 3]
```diff
@@ -247,6 +275,88 @@ beam_manager_impl::init_worker(
...
+            {
+                std::shared_ptr<beam_rpc_message> currentMessage = m_pendingSend.front();
+                m_pendingSend.pop();
+                send_message(currentMessage);
+            }
+        }
+
+        {
+            static const int numMessageRetries = 10;
+            static const std::chrono::milliseconds messageRetryInterval = std::chrono::milliseconds(10 * 1000); // 10-second retry interval
```
Comment (jicailiu) [inline Source/Services/Beam/beam_manager_impl.cpp]: Nit: why don't you use chrono::seconds

Comment (edjamesmsft) [inline Source/Services/Beam/beam_manager_impl.cpp]: we use milliseconds for the timestamps - easier to use milliseconds here than awkward conversion logic below

---
In reply to: [117138035](https://github.com/WatchBeam/interactive-cpp/pull/6#discussion_r117138035) [](ancestors = 117138035)

File: Source/Services/Beam/beam_manager_impl.cpp (hunk 2 above)
Comment (edjamesmsft) [inline Source/Services/Beam/beam_manager_impl.cpp]: ah, good catch. I'll do a pass through on another change to update throughout.

---
In reply to: [117137492](https://github.com/WatchBeam/interactive-cpp/pull/6#discussion_r117137492) [](ancestors = 117137492)

File: Source/Services/Beam/beam_manager_impl.cpp [hunk 4]
```diff
@@ -109,38 +114,55 @@ beam_manager_impl::initialize(
...
+    }
 
     // Create long-running message processing task
-    m_processMessagesTask = pplx::create_task([thisWeakPtr]()
+    if (false == m_processing)
     {
-        std::shared_ptr<beam_manager_impl> pThis;
-        pThis = thisWeakPtr.lock();
-        if (nullptr != pThis)
+        m_processMessagesTask = pplx::create_task([thisWeakPtr]()
```
Comment (adityatMS) [inline Source/Services/Beam/beam_manager_impl.cpp]: >m_processMessagesTask [](start = 8, length = 21)

Dont need this as you aren't really using this anywhere else.

**Comment by adityatMS** [comment]
> beam_manager_impl::initialize(

Add a lock to protect against multiple threads initializing this.

---
Refers to: Source/Services/Beam/beam_manager_impl.cpp:112 in 349f316. [](commit_id = 349f3169b439793f5c27740216020ca99951026a, deletion_comment = False)

**Comment by adityatMS** [comment]
>         // Only one user supported at this time, but allow over-writing of the current user

I would recommend returning an error instead.

---
Refers to: Source/Services/Beam/beam_manager_impl.cpp:183 in 349f316. [](commit_id = 349f3169b439793f5c27740216020ca99951026a, deletion_comment = False)

File: Source/Services/Beam/beam_manager_impl.cpp (hunk 3 above)
Comment (adityatMS) [inline Source/Services/Beam/beam_manager_impl.cpp]: Would recommend defining all statics at the top of the file.

**Pull request merged**